Do not return error on garbage data after last Ogg page

AlexTMjugador opened the pull request:
Some admittedly badly-encoded Ogg Vorbis files in the wild contain trailing garbage data after the last Ogg page, such as this one: https://freesound.org/people/stomachache/sounds/157587/

The page reading code explicitly handled this case by returning an error, which makes sense to do given that the Ogg specification is not clear on how this condition should be dealt with. Moreover, programs like `ogginfo` also show warnings for these kind of files. However, being this strict causes codec crates that depend on this one to not be able to play back such files properly, as discussed in https://github.com/RustAudio/ogg/issues/7.

Fix the situation by being more lenient with this nonsense. Some additional status information is kept to distinguish this case from empty files.

I've not tested the seeking code thoroughly after these changes, but both empty, normal and normal but with trailing garbage files seem to work fine. Tests also pass.

AlexTMjugador committed:
Do not return error on garbage data after last Ogg page

Some admittely badly-encoded Ogg Vorbis files in the wild contain
trailing garbage data after the last Ogg page, such as this one:
https://freesound.org/people/stomachache/sounds/157587/.

The page reading code explicitly handled this case by returning an
error, which makes sense to do given that the Ogg specification is not
clear on how this condition should be dealt with. Moreover, programs
like `ogginfo` also show warnings for these kind of files. However,
being this strict causes codec crates that depend on this one to not be
able to play back such files properly, as discussed in
https://github.com/RustAudio/ogg/issues/7.

Fix the situation by being more lenient with this nonsense. Some
additional status information is kept to distinguish this case from
empty files.

est31 commented:
Thanks. I wonder if one could add a test?

AlexTMjugador commented:
It could do that, but I'm not sure what kind of inputs could cause the changes to exhibit bad behavior. Do you have any ideas? :smile:

AlexTMjugador committed:
Tweak comment to make it clearer

est31 commented:
I was wondering about a test that includes garbage after the last page, and the test ensuring that the garbage does not yield an error.

AlexTMjugador committed:
Add unit test for proper trailing data after last page handling

AlexTMjugador commented:
I've just added a unit test for that :+1:

AlexTMjugador committed:
Ensure that trailing data is not returned in a packet in unit test

I don't expect this additional check to be useful in practice, but it is
an invariant that should be held in the test case after all.

est31 reviewed (approved):
Thanks!
